(Samples) Transfer in batch 2 (#235)
OfficeDev/office-scripts-docs: docs/resources/samples/filter-table-get-visible-range.md

> ---

Reviewer: After reading through the whole sample, I realize this creates a Map from an Excel table, which seems both complex and worthy of up-front mention.

Reviewer: Perhaps the sample could be remade into two parts: one to show this Table -> Map conversion and one to show applying that logic to a filtered table about stations.

Author: Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)

title: 'Filter Excel table and get visible range'
description: 'Learn how to use Office Scripts to filter an Excel table and get the visible range as an array of objects.'
ms.date: 03/04/2021
localization_priority: Normal
---

> # Filter Excel table and get visible range as array of objects

Reviewer: I think highlighting JSON as the return type has discoverability benefits.
```suggestion
# Filter Excel table and get visible range as a JSON object
```

> Shows how to use Office Scripts to filter an Excel table and get the visible range as an array of objects

Reviewer: Suggestion to make this more scenario driven.
```suggestion
This sample filters an Excel table and returns the visible range as a JSON object. This JSON could be provided to a Power Automate flow as part of a larger solution.
```

## Example scenario

> * Apply filter on a column of a table.

Reviewer: ```suggestion
* Apply a filter to a table column.
```

> * Extract visible range after filtering and extract data as array of objects.

Reviewer: ```suggestion
* Extract the visible range after filtering.
```

> * Assemble an object with following JSON structure and return for a future action to use.

Reviewer: ```suggestion
* Assemble and return an object with a specific JSON structure.
```

### Sample JSON

Each key represents a unique value of a table. Each array instance represents the row that is visible when the corresponding filter is applied.

```json
{
  "Station-1": [{
    "Station": "Station-1",
    "Shift": "Morning",
    "Date": "27-Oct-20",
    "Responsible": "Debra Berger",
    "Reason": ""
  }, {
    "Station": "Station-1",
    "Shift": "Afternoon",
    "Date": "27-Oct-20",
    "Responsible": "Delia Dennis",
    "Reason": ""

>   }, {

Reviewer: Do we need all of this? Can we just list two then go "..."?

    "Station": "Station-1",
    "Shift": "Afternoon",
    "Date": "28-Oct-20",
    "Responsible": "Lidia Holloway",
    "Reason": ""
  }],
  "Station-2": [{
    "Station": "Station-2",
    "Shift": "Morning",
    "Date": "27-Oct-20",
    "Responsible": "Gerhart Moller",
    "Reason": ""
  }, {
    "Station": "Station-2",
    "Shift": "Afternoon",
    "Date": "28-Oct-20",
    "Responsible": "Grady Archie",
    "Reason": ""
  }],
  "Station-3": [{
    "Station": "Station-3",
    "Shift": "Morning",
    "Date": "27-Oct-20",
    "Responsible": "Isaiah Langer",
    "Reason": ""
  }]
}
```

> ## Solution

Reviewer: This section seems unnecessary.

The solution has a single Office Script:

* Performs transformation as described previously and extracts required data

## Office Script

The following script filters a table and gets the visible range.

Download the sample file <a href="table-filter.xlsx">table-filter.xlsx</a> and use it with this script to try it out yourself!

```ts
function main(workbook: ExcelScript.Workbook): ReturnTemplate {
  const table1 = workbook.getTable("Table1");
  const keyColumnValues: string [] = table1.getColumnByName('Station').getRangeBetweenHeaderAndTotal().getValues().map(v => v[0] as string);

>   // const uniqueKeys= [...Array.from(new Set(keyColumnValues))];

Reviewer: Let's remove commented-out code.

>   const uniqueKeys = keyColumnValues.filter((v, i, a) => a.indexOf(v) === i);

Reviewer: This map building code is super dense. It would greatly benefit from being split into several lines and having comments.

Author: Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)

  console.log(uniqueKeys);
  const returnObj: ReturnTemplate = {}

  uniqueKeys.forEach((key: string) => {
    table1.getColumnByName('Station').getFilter()
      .applyValuesFilter([key]);
    const rangeView = table1.getRange().getVisibleView();
    returnObj[key] = returnObjectFromValues(rangeView.getValues() as string[][]);
  })
  table1.getColumnByName('Station').getFilter().clear();
  console.log(JSON.stringify(returnObj));
  return returnObj
}

function returnObjectFromValues(values: string[][]): BasicObj[] {
  let objArray = [];
  let objKeys: string[] = [];
  for (let i=0; i < values.length; i++) {
    if (i===0) {
      objKeys = values[i]
      continue;
    }
    let obj = {}
    for (let j=0; j < values[i].length; j++) {
      obj[objKeys[j]] = values[i][j]
    }
    objArray.push(obj);
  }
  return objArray;
}

interface BasicObj {
  [key: string] : string
}

interface ReturnTemplate {
  [key: string]: BasicObj[]
}

> ```

Reviewer: **Edit:** Wrote this before my comment about making this a generic Map sample.
Rather than having generic names, I think using terms specific to the scenario would help increase comprehension. Maybe StationList and Station?

Author: Tracking with [feature](https://office.visualstudio.com/OC/_workitems/edit/4924924)
